Sources/Rendering/OpenGL/CustomPostProcessingPass/index.js

```diff
@@ -0,0 +1,317 @@
+    return weight <= 0 ? 1 : weight;
+  };
+
+  // handles post-processing via convolution kernel and call delegate
+  publicAPI.traverse = (viewNode, parent = null) => {
+    if (model.deleted) {
+      return;
+    }
+
+    // check if kernel dimension is valid
+    if (model.kernelDimension % 2 !== 1) {
+      vtkErrorMacro(
+        'Invalid kernel dimension! Kernel dimension must be odd (e.g. 3, 5, 7, ...).'
+      );
+      return;
+    }
+
+    // if no kernel is set, use the default kernel (no post-processing)
+    if (model.kernel === null) {
+      model.kernel = new Float32Array(model.kernelDimension);
+      model.kernel[Math.floor(model.kernelDimension / 2)] = 1;
+    }
+
+    const kernelLength = model.kernelDimension * model.kernelDimension;
+    if (model.kernel.length !== kernelLength) {
+      vtkErrorMacro(
+        `The given kernel is invalid. 2D convolution kernels have to be 1D arrays with ${kernelLength} components representing the ${
+          model.kernelDimension
+        }x${model.kernelDimension} kernel in row-major form.`
+      );
+      return;
+    }
+
+    // prepare framebuffer // allocate framebuffer if needed and bind it
+    if (model.framebuffer === null) {
+      model.framebuffer = vtkOpenGLFramebuffer.newInstance();
+    }
+
+    const size = viewNode.getSize();
+    const gl = viewNode.getContext();
+
+    if (gl === null) {
+      // nothing to do -> no render context
+      // traverse delegate passes -> has to be done in order for the vtk render-pipeline to work correctly
+      model.delegates.forEach((val) => {
+        val.traverse(viewNode, publicAPI);
+      });
+      return;
+    }
+
+    // prepare rendering
+    if (model.VBOBuildTime.getMTime() < publicAPI.getMTime()) {
+      model.tris.setOpenGLRenderWindow(viewNode);
+      publicAPI.buildVertexBuffer();
+    }
+
+    // store framebuffer bindings to restore them later
+    model.framebuffer.setOpenGLRenderWindow(viewNode);
+    model.framebuffer.saveCurrentBindingsAndBuffers();
+
+    const fbSize = model.framebuffer.getSize();
+
+    if (fbSize === null || fbSize[0] !== size[0] || fbSize[1] !== size[1]) {
+      // create post-processing framebuffer if not already existing
+      model.framebuffer.create(size[0], size[1]);
+      model.framebuffer.populateFramebuffer();
+    }
+
+    // bind framebuffer to re-direct the render-output of the delegate passes to the buffer
+    model.framebuffer.bind();
+
+    // do the delegate rendering
+    model.delegates.forEach((val) => {
+      val.traverse(viewNode, publicAPI);
+    });
+
+    // now draw the distorted values
+    model.framebuffer.restorePreviousBindingsAndBuffers();
+
+    // check if kernel dimension has changed and convolution shader needs to be re-compiled
+    if (
+      model.convolutionShader !== null &&
+      model.oldKernelDimension !== model.kernelDimension
+    ) {
+      // model.convolutionShader.cleanup(); // todo: what does this do?
```

Comment:
I think you can uncomment this (it just calls `gl.deleteShader`)

Reply:
We used to have this active but it caused crashes when changing the kernel size. Do you think we should try to re-enable it or just remove the line entirely?

Reply:
I think it would be interesting to see if 
```
  model.convolutionShader.release();
  viewNode.getShaderCache().setLastShaderBound(null);
  model.convolutionShader.cleanup();
```
Fixes the crash issue.  

If that doesn't work, then you can drop the line.

Reply:
We tried it with your suggestion but it still crashes. In the most recent commit we omitted the line.

![image](https://user-images.githubusercontent.com/68997390/89035052-8a4baa80-d33a-11ea-9010-3c769083dfce.png)

Reply:
It's ok to just drop that line.